ctapipe/image/cleaning.py

```diff
@@ -356,3 +356,26 @@ def fact_image_cleaning(
         geom, pixels_to_keep, arrival_times, min_number_neighbors, time_limit
     )
     return pixels_to_keep
+
+
+def largest_island(islands_labels):
+    """Find the biggest island and filter it from the image.
+
+    This function takes a list of islands in an image and isolates the largest one
+    for later parametrization.
+
+    Parameters
+    ----------
+
+    islands_labels : array
+        Flattened array containing a list of labelled islands from a cleaned image.
+        Second returned value of the function 'number_of_islands'.
+
+    Returns
+    -------
+
+    islands_labels : array
+        A boolean mask created from the input labels and filtered for the largest island.
+
+    """
```

Comment:
See cta-observatory/protopipe/issue/15, check if there are even islands, e.g.:

```
valid = island_labels > 0
if np.count_nonzero(valid) == 1:
     return np.zeros(island_labels.shape, dtype='bool')
```
```

Reply:
Ok, I understand what you mean.
Still, I would like to elaborate a bit on this.

`largest_island` depends on `number_of_islands`: in principle, there is no reason to call it if there are no islands.

Do you think it is more efficient to check for the existence of the islands inside `largest_island` or outside by using the already computed number by your function? In the end, both solutions use a conditional, am I right?

Also, there is no reason to call `largest_island` in case of only one island surviving the cleaning, since the original cleaning mask has already only one island.

The way I fixed the bug in protopipe follows this principle by using the ctapipe functions in a modular way: first I call `number_of_islands` and then I check how many there are and I call `largest_island` only if necessary.

Naturally, if we find a more efficient way to do it even better: I will call the ctapipe functions anyway once the new stable release is out.

Reply:
I for myself doubt that it is even a good idea to calculate hillas parameters only on the largest island.

Do you have comparisons?

Reply:
Anyways, this function should either return an empty mask if no island is there or it should raise a meaningfull error message. The error message from deep in numpy not indication the real problem is not helpful for someone using this function.

Reply:
> I for myself doubt that it is even a good idea to calculate hillas parameters only on the largest island.
> 
> Do you have comparisons?

This is a very good question! The work we are doing now with CTA-MARS and in the near future should shed some light on this.

Reply:
> Anyways, this function should either return an empty mask if no island is there or it should raise a meaningfull error message. The error message from deep in numpy not indication the real problem is not helpful for someone using this function.

Ok, I get it.
I'll will improve it to catch the case with no islands and the case with one should be automatic.

Thanks for the input!

Reply:
> I for myself doubt that it is even a good idea to calculate hillas parameters only on the largest island.
> 
> Do you have comparisons?

It should be for the direction reconstruction only.
Don't do it for the energy or classification.

Reply:
In your implementation, why did you use a boolean instead of using directly the labels array?

Meaning, why didn't you just do the following?

```
if np.count_nonzero(island_labels) == 0:
    return np.zeros(island_labels.shape, dtype='bool')
```

Reply:
> > I for myself doubt that it is even a good idea to calculate hillas parameters only on the largest island.
> > Do you have comparisons?
> 
> It should be for the direction reconstruction only.
> Don't do it for the energy or classification.

Yes, this WOULD be the case for protopipe.
For energy estimation and g/h classification, we should use the content of the whole set of islands, but for the moment the cleaning is the same.
The reason is that the performance of these two cleanings has not been studied in model estimation with protopipe.
I hope that in the future this part will be studied in more detial, but for the moment this is what we have.

Reply:
> Meaning, why didn't you just do the following?

Did not think about it. Should be fine!